### ISSUE TITLE
'MODULE FAILURE' message when notification of password expiration

### ISSUE THREAD

**Issue by JimFicarra** [ansible/ansible#15601]
This was originally submitted as https://github.com/ansible/ansible-modules-core/issues/3439 which has been closed.  @jctanner requested I post this issue in this repo instead.
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

setup, file, likely others (those were the two that I tested)
##### ANSIBLE VERSION

```
ansible 2.0.0.2
Tower 2.4.4
```
##### CONFIGURATION

pipelining=true
timeout=30
##### OS / ENVIRONMENT

Control Host: CentOS Linux release 7.0.1406 (Core)
Target Host: Centos 6.3
##### SUMMARY

When the password of the remote user you are connecting to has a password that is going to expire, you receive a non-descript MODULE FAILURE error

**_Password expiration notice triggering this:_**
Warning: Your password will expire in 5 days on Tue 19 Apr 2016 06:58:59 PM UTC
##### STEPS TO REPRODUCE
- Use an account with a password that is about to expire. In our case, the run was to sudo as root, but not sure that makes a difference.
- Execute a playbook against a target host with that remote user

```
- hosts: webservers
  become: yes

  gather_facts: true


  tasks:
    - name: create dir as someone
      file:
        path: /tmp/water
        state: directory

```

##### EXPECTED RESULTS
- /tmp/water created as root (assuming user password is not actually expired yet and user has sudo as root rights, which the user did)
- I would expect the playbook to fail if the password had actually expired, but in this case it was simply a warning and the play should have succeeded
- I would also expect a more descriptive error message.  MODULE FAILURE is too generic and did not point to a specific cause

When we did the password change, the error went away.
##### ACTUAL RESULTS

Renamed some specifics to protect the innocent

```
Using /etc/ansible/ansible.cfg as config file
SSH password: 
SUDO password[defaults to SSH password]: 
1 plays in test_download.yml

PLAY ***************************************************************************

TASK [Ensure /tmp/water folder exists] ********************************************
task path: /var/lib/awx/projects/_1063__application_engineering/test_download.yml:8
<server1.mycompany.com> ESTABLISH SSH CONNECTION FOR USER: userid123
<server2.mycompany.com> ESTABLISH SSH CONNECTION FOR USER: userid123
<server2.mycompany.com> SSH: EXEC sshpass -d25 ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o User=userid123 -o ConnectTimeout=30 -o ControlPath=/tmp/ansible_tower_vkmTPd/cp/ansible-ssh-%h-%p-%r server2.mycompany.com '/bin/sh -c '"'"'sudo -H -S  -p "[sudo via ansible, key=ciwwtampnqikadsfwwkxwszmozbasafv] password: " -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-ciwwtampnqikadsfwwkxwszmozbasafv; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python'"'"'"'"'"'"'"'"''"'"''
<server1.mycompany.com> SSH: EXEC sshpass -d24 ssh -C -q -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o User=userid123 -o ConnectTimeout=30 -o ControlPath=/tmp/ansible_tower_vkmTPd/cp/ansible-ssh-%h-%p-%r server1.mycompany.com '/bin/sh -c '"'"'sudo -H -S  -p "[sudo via ansible, key=guaywsgdbbtlrxouxrraebsustvesxbk] password: " -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-guaywsgdbbtlrxouxrraebsustvesxbk; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python'"'"'"'"'"'"'"'"''"'"''
fatal: [server2.mycompany.com]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "file"}, "module_stderr": "", "module_stdout": "{\"group\": \"root\", \"uid\": 0, \"changed\": true, \"owner\": \"root\", \"state\": \"directory\", \"gid\": 0, \"mode\": \"0755\", \"invocation\": {\"module_args\": {\"directory_mode\": null, \"force\": false, \"remote_src\": null, \"path\": \"/tmp/water/\", \"owner\": null, \"follow\": false, \"group\": null, \"state\": \"directory\", \"content\": null, \"serole\": null, \"diff_peek\": null, \"setype\": null, \"selevel\": null, \"original_basename\": null, \"regexp\": null, \"validate\": null, \"src\": null, \"seuser\": null, \"recurse\": false, \"delimiter\": null, \"mode\": null, \"backup\": null}}, \"path\": \"/tmp/water/\", \"size\": 4096}\nWarning: Your password will expire in 5 days on Tue 19 Apr 2016 06:58:59 PM UTC\n", "msg": "MODULE FAILURE", "parsed": false}
fatal: [server1.mycompany.com]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "file"}, "module_stderr": "", "module_stdout": "{\"group\": \"root\", \"uid\": 0, \"changed\": true, \"owner\": \"root\", \"state\": \"directory\", \"gid\": 0, \"mode\": \"0755\", \"invocation\": {\"module_args\": {\"directory_mode\": null, \"force\": false, \"remote_src\": null, \"path\": \"/tmp/water/\", \"owner\": null, \"follow\": false, \"group\": null, \"state\": \"directory\", \"content\": null, \"serole\": null, \"diff_peek\": null, \"setype\": null, \"selevel\": null, \"original_basename\": null, \"regexp\": null, \"validate\": null, \"src\": null, \"seuser\": null, \"recurse\": false, \"delimiter\": null, \"mode\": null, \"backup\": null}}, \"path\": \"/tmp/water/\", \"size\": 4096}\nWarning: Your password will expire in 5 days on Tue 19 Apr 2016 06:58:59 PM UTC\n", "msg": "MODULE FAILURE", "parsed": false}

PLAY RECAP *********************************************************************
server1.mycompany.com : ok=0    changed=0    unreachable=0    failed=1   
server2.mycompany.com : ok=0    changed=0    unreachable=0    failed=1  
```

**Comment by abadger** [comment]
I just tested this and it seems to be a variant on this: http://docs.ansible.com/ansible/intro_configuration.html#pipelining

You can test again with either requiretty turned off in /etc/sudoers or pipelining off to verify whether we're seeing the same thing.

i don't think this is fixable but I'll see if there's a way to better detect the error and give a better message (it might not be possible... It all depends on what information we get back from ssh in this circumstance).

**Comment by abadger** [comment]
Hmm... so thinking about this more, i'm not sure whether this is the same problem or not.  the output I get is different from yours when I have pipelining and requiretty.  So the question is if you turn pipelining off, do things start working?

I'm unable to replicate your error condition (getting the same output as you do).  Everything seems to work fine unless I have the requiretty and pipelining combo (which gives a different error).

Manual login looks like this:

```
$ ssh rhel6                                                     (15:12:05)
Warning: your password will expire in 98 days
Last login: Wed Apr 27 14:44:25 2016 from 192.168.122.1
$ chage -l badger                                                                          (15:12:32)
Last password change                                    : Jan 01, 2015
Password expires                                        : Aug 03, 2016
Password inactive                                       : never
Account expires                                         : Jan 20, 11761191
Minimum number of days between password change          : 0
Maximum number of days between password change          : 580
Number of days of warning before password expires       : 100
```

Working Ansible run:

```
$ ansible-playbook 15601.yml -i inven -K                        (15:14:18)
SUDO password: 

PLAY [pass-test] ***************************************************************

TASK [create dir as someone] ***************************************************
changed: [pass-test]

PLAY RECAP *********************************************************************
pass-test                  : ok=1    changed=1    unreachable=0    failed=0
$ ansible --version                                             (15:17:34)
ansible 2.0.0.2 (detached HEAD 7de237c5a1) last updated 2016/04/27 15:17:29 (GMT -700)
  lib/ansible/modules/core: (stable-2.0 4b42c6ad59) last updated 2016/04/26 10:34:02 (GMT -700)
  lib/ansible/modules/extras: (stable-2.0 4eb177e545) last updated 2016/04/26 10:34:17 (GMT -700)
  config file = /srv/ansible/troubleshooting/ansible.cfg
  configured module search path = Default w/o overrides
```

Is there anything else you can tell me about your setup that might help to reproduce this?

**Comment by JimFicarra** [comment]
Pipelining was on and the hosts did not have requiretty set.  Are you just testing with ansible core or are you throwing Tower into the mix? This was occurring when running in Tower.

**Comment by abadger** [comment]
I was just testing with ansible core.  I don't think tower will make a difference.  But thinking about how you could get the warning after the module's output, maybe there's pam or syslog configuration that causes this.  Do you happen to know if you're changing RHEL/Centos6's defaults for either of those in some way?  I'd like to be able to reproduce so that I know when I've fixed this.

Failing that, I think I can propose fixes for you to test but that's obviously more work for you that I'd like to avoid.

**Comment by abadger** [comment]
Ah ha!  @alikins thinks it might be that using krb5 might cause the Warning message to be printed to stdout differently so that could create the case where the module prints the json results before the warning message is emitted.

**Comment by abadger** [comment]
Okay, I wasn't able to test but judging from the output from your ansible run, that pull request should fix this problem.  It strips away text that comes after the json string that the module returned.  Give it a shot and see what happens.  I've merged it to the devel and stable-2.1 branches so it should be in the 2.1.0 release which is coming soon.  (We might not see comments to closed issues so if it still gives you problems, please create a new issue and link to this bug report in the new one.)

Thanks for the help debugging this!

**Comment by JimFicarra** [comment]
Thanks all - I reproduced the conditions and validated that the fix works in Ansible 2.1.0.0 and Ansible 2.1.1.0
